lib/saml/response.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+module SAML
+  class Response < OneLogin::RubySaml::Response
+    ERRORS = { clicked_deny:   { code: '001',
+                                 tag: :clicked_deny,
+                                 short_message: 'Subject did not consent to attribute release',
+                                 level: :warn },
+               auth_too_late:  { code: '002',
+                                 tag: :auth_too_late,
+                                 short_message: 'Current time is on or after NotOnOrAfter condition',
+                                 level: :warn },
+               auth_too_early: { code: '003',
+                                 tag: :auth_too_early,
+                                 short_message: 'Current time is earlier than NotBefore condition',
+                                 level: :error },
+               blank:           { code: '007',
+                                  tag: :blank,
+                                  short_message: 'Blank response',
+                                  level: :error },
+               unknown:         { code: '007',
+                                  tag: :unknown,
+                                  short_message: 'Other SAML Response Error(s)',
+                                  level: :error } }.freeze
+
+    def normalized_errors
+      @normalized_errors ||= []
+    end
+
+    def valid?
+      @normalized_errors = []
+      # passing true collects all validation errors
+      is_valid_result = is_valid?(true)
```

Comment:
hmmm, i'm wondering if we can just add our errors to the same errors array that saml_response has or do you think there are consequences with doing that? I'll have to review this more thoroughly after looking at SamlResponse code in OneLogin... This is why this is taking so long for me to review.

Reply:
in other words, override is_valid? (first calling super maybe) then adding our errors in as well. And aliasing is_valid? to valid? as well because is_valid? is stilly.

Reply:
That's a possibility, but it seems we want all of the different attributes from the ERRORS hash to be used later. It didn't seem like the right pace to make the statsd/sentry calls though, since it's reasonable to call valid? more than once.